src/dag_combine.rs

```diff
@@ -0,0 +1,56 @@
+use digraph::PyDiGraph;
+use pyo3::prelude::*;
+use petgraph::algo;
+use pyo3::Python;
+use pyo3::types::PyAny;
+use std::cmp::Ordering;
+
+fn contains_weight(py: Python, dag: &PyDiGraph, obj: &PyObject) -> i32 {
+    let mut index :i32 = -1;
+
+    for node in dag.graph.node_indices() {
+        let weight = dag.graph.node_weight(node).unwrap();
+        if ! weight.is_none(py) {
+
+            let predicate_callable = |a: &PyAny, b: &PyAny| -> PyResult<bool> {
+                let res = a.compare(b)?;
+                Ok(res == Ordering::Equal)
+            };
+
+            if predicate_callable(obj.as_ref(py), weight.as_ref(py)).unwrap() {
+                index = node.index() as i32;
+                break;
+            }
+        }
+    }
+    return index;
+}
+
+pub fn combine_graphs(
+    py: Python,
+    dag0: &PyDiGraph,
+    dag1: &PyDiGraph,
+) -> PyResult<PyDiGraph> {
+    let first = &dag0.graph;
+    let second = &dag1.graph;
+    let mut combined = PyDiGraph {
+        graph: first.clone(),
+        cycle_state: algo::DfsSpace::default(),
+        check_cycle: false,
+        node_removed: false,
+    };
+    for node in second.node_indices(){
+        let weight = &second[node].clone();
+        if ! weight.is_none(py) {
+            let index = contains_weight(py, dag0, weight);
+            if index != -1 {
+                let other_node = combined.add_node(second[node].clone()).unwrap();
+                combined.add_edge(index as usize, other_node, py.None())?;
```

Comment:
Won't this edge just get deleted (and add a self edge to `other_node`) by the `merge_node` call below? We know that the weights are the same here so the `merge_node` call will remove `index` from the graph which will drop any edges on `index`. Or am I missing something?